One Shot Single - using Podman > PostgresInitScriptCopyFailedSoloError invalid container path /tmp

## Failure Summary

The One Shot Single using Podman job failed on branch trunk-merge slash pr-5534 slash 8d587ca7-96f6-483e-846b-36321833d866. Error cause is Postgres init script copy failure due to invalid container path.

## Error Details

    OneShotDeployFailedSoloError: Deploy failed: Failed to copy Mirror Node Postgres initialization script to container in namespace one-shot: Invalid container path in destination: /tmp. Rollback skipped (--no-rollback).
      at DefaultOneShotCommand.performRollback (/home/runner/work/solo/solo/src/commands/one-shot/default-one-shot.ts:211:13)
      at DefaultOneShotCommand.deployInternal (/home/runner/work/solo/solo/src/commands/one-shot/default-one-shot.ts:284:18)
      at process.processTicksAndRejections (node:internal/process/task_queues:103:5)
      at async Object.handler (/home/runner/work/solo/solo/src/core/command-path-builders/command-builder.ts:141:47)
    Caused by: PostgresInitScriptCopyFailedSoloError: Failed to copy Mirror Node Postgres initialization script to container in namespace one-shot: Invalid container path in destination: /tmp
      at PostgresSharedResource.initializeMirrorNode (/home/runner/work/solo/solo/src/core/shared-resources/postgres.ts:142:13)
      at process.processTicksAndRejections (node:internal/process/task_queues:103:5)
      at async Task.task (/home/runner/work/solo/solo/src/commands/mirror-node.ts:980:9)
      at async Task.run (/home/runner/work/solo/solo/node_modules/listr2/dist/index.mjs:1970:5)
    Caused by: KubeContainerInvalidPathError: Invalid container path in destination: /tmp
      at K8ClientContainer.copyTo (/home/runner/work/solo/solo/src/integration/kube/k8-client/resources/container/k8-client-container.ts:248:13)
      at process.processTicksAndRejections (node:internal/process/task_queues:103:5)
      at async PostgresSharedResource.initializeMirrorNode (/home/runner/work/solo/solo/src/core/shared-resources/postgres.ts:139:7)
      at async Task.task (/home/runner/work/solo/solo/src/commands/mirror-node.ts:980:9)
      at async Task.run (/home/runner/work/solo/solo/node_modules/listr2/dist/index.mjs:1970:5)

## Workflow Information

Run: https://github.com/hiero-ledger/solo/actions/runs/30798829326
Job: https://github.com/hiero-ledger/solo/actions/runs/30798829326/job/91641873169
Issue: https://github.com/hiero-ledger/solo/issues/5578

Another incident of podman failure, not sure correlated
https://github.com/hiero-ledger/solo/actions/runs/30820669705/job/91713076597?pr=5363

```
  1 failing

  1) One Shot Single E2E Test Suite
       One Shot Single E2E Test
         one-shot-single-minimal: deploy:
     OneShotDeployFailedSoloError: Deploy failed: Data validation failed: Mirror Node in remote config with id 1 was not found in namespace: one-shot, cluster: one-shot,    labels: app.kubernetes.io/name=importer,app.kubernetes.io/component=importer,app.kubernetes.io/instance=mirror-1 (expected: {"metadata":{"id":1,"namespace":"one-shot","cluster":"one-shot","phase":"deployed"}}, found: {}). Rollback skipped (--no-rollback).
      at DefaultOneShotCommand.performRollback (/home/runner/work/solo/solo/src/commands/one-shot/default-one-shot.ts:211:13)
      at DefaultOneShotCommand.deployInternal (/home/runner/work/solo/solo/src/commands/one-shot/default-one-shot.ts:284:18)
      at process.processTicksAndRejections (node:internal/process/task_queues:103:5)
      at async Object.handler (/home/runner/work/solo/solo/src/core/command-path-builders/command-builder.ts:141:47)
     Caused by: DataValidationError: Data validation failed: Mirror Node in remote config with id 1 was not found in namespace: one-shot, cluster: one-shot,    labels: app.kubernetes.io/name=importer,app.kubernetes.io/component=importer,app.kubernetes.io/instance=mirror-1 (expected: {"metadata":{"id":1,"namespace":"one-shot","cluster":"one-shot","phase":"deployed"}}, found: {})
      at Function.buildValidationError (/home/runner/work/solo/solo/src/core/config/remote/remote-config-validator.ts:191:12)
      at <anonymous> (/home/runner/work/solo/solo/src/core/config/remote/remote-config-validator.ts:172:37)
      at process.processTicksAndRejections (node:internal/process/task_queues:103:5)
      at async Promise.all (index 1)
      at async RemoteConfigValidator.validateComponents (/home/runner/work/solo/solo/src/core/config/remote/remote-config-validator.ts:121:5)
      at async RemoteConfigRuntimeState.loadAndValidate (/home/runner/work/solo/solo/src/business/runtime-state/config/remote/remote-config-runtime-state.ts:403:5)
      at async Task.task (/home/runner/work/solo/solo/src/commands/explorer.ts:694:13)
      at async Task.run (/home/runner/work/solo/solo/node_modules/listr2/dist/index.mjs:1970:5)
     Caused by: 

```

--- a/src/integration/kube/k8-client/resources/container/k8-client-container.ts
+++ b/src/integration/kube/k8-client/resources/container/k8-client-container.ts
@@ -345,7 +345,7 @@ export class K8ClientContainer implements Container {
     // Retry transient exec failures that occur before the command runs: kubelet reporting "container not found" during rolling restarts, and API server to kubelet tunnel errors (connection upgrade, dial, timeout).
     const maxAttempts: number = 30;
     const retryableTransientFailure: RegExp =
-      /(container not found|unable to upgrade connection|internal error occurred: timeout occurred|error dialing backend)/i;
+      /(container not found|unable to upgrade connection|internal error occurred: timeout occurred|error dialing backend|connection reset by peer)/i;
 
     for (let attempt: number = 1; attempt <= maxAttempts; attempt++) {
       try {
@@ -368,18 +368,42 @@ export class K8ClientContainer implements Container {
   }
 
   public async hasDir(destinationPath: string): Promise<boolean> {
+    const maxAttempts: number = 3;
+
+    for (let attempt: number = 1; attempt <= maxAttempts; attempt++) {
+      const result: string = await this.probeDirectory(destinationPath);
+
+      if (result === 'true' || result === 'false') {
+        return result === 'true';
+      }
+
+      // an exec stream cut mid-flight (e.g. containerd restart) can exit 0 with no output, so only the literal probe answers are trusted
+      this.logger.warn(
+        `hasDir: unexpected probe output '${result}' on attempt ${attempt} of ${maxAttempts} [podName: ${this.containerReference.parentReference.name} -c ${this.containerReference.name}, path: ${destinationPath}]`,
+      );
+
+      if (attempt < maxAttempts) {
+        await sleep(Duration.ofMillis(attempt * 1000));
+      }
+    }
+
+    throw new KubeContainerOperationFailedError(
+      `hasDir ${destinationPath}`,
+      new Error('directory probe returned no usable output after retries'),
+    );
+  }
+
+  private async probeDirectory(destinationPath: string): Promise<string> {
     const bashScript: string = `[[ -d "${destinationPath}" ]] && echo -n "true" || echo -n "false"`;
     try {
-      const result: string = await this.execContainer(['bash', '-c', bashScript]);
-      return result === 'true';
+      return await this.execContainer(['bash', '-c', bashScript]);
     } catch (error) {
       this.logger.debug(
         `hasDir failed using bash for ${this.containerReference.parentReference.name}:${this.containerReference.name}, retrying with /bin/sh`,
         error,
       );
       const shScript: string = `[ -d "${destinationPath}" ] && echo -n "true" || echo -n "false"`;
-      const result: string = await this.execContainer(['/bin/sh', '-c', shScript]);
-      return result === 'true';
+      return await this.execContainer(['/bin/sh', '-c', shScript]);
     }
   }
 

--- a/test/unit/integration/kube/k8-client-container.test.ts
+++ b/test/unit/integration/kube/k8-client-container.test.ts
@@ -49,6 +49,7 @@ describe('K8ClientContainer execContainer', (): void => {
     'error: Internal error occurred: Timeout occurred',
     'Error from server: error dialing backend: dial tcp 10.89.0.2:10250: connect: connection refused',
     'error: unable to upgrade connection: container not found ("postgresql")',
+    'error: error reading from server: read unix @->/run/containerd/containerd.sock: read: connection reset by peer',
   ]) {
     it(`retries and succeeds after transient failure: ${stderr}`, async (): Promise<void> => {
       execKubectlStub.onFirstCall().rejects(kubectlFailure(stderr));
@@ -75,5 +76,57 @@ describe('K8ClientContainer execContainer', (): void => {
     }
 
     expect(execKubectlStub).to.have.been.calledOnce;
+  });
+
+  describe('hasDir', (): void => {
+    it('returns true when the probe prints true', async (): Promise<void> => {
+      execKubectlStub.resolves('true');
+
+      expect(await containerClient.hasDir('/tmp')).to.be.true;
+      expect(execKubectlStub).to.have.been.calledOnce;
+    });
+
+    it('returns false when the probe prints false', async (): Promise<void> => {
+      execKubectlStub.resolves('false');
+
+      expect(await containerClient.hasDir('/missing')).to.be.false;
+      expect(execKubectlStub).to.have.been.calledOnce;
+    });
+
+    it('retries when an interrupted exec returns no output, then succeeds', async (): Promise<void> => {
+      execKubectlStub.onFirstCall().resolves('');
+      execKubectlStub.onSecondCall().resolves('true');
+
+      expect(await containerClient.hasDir('/tmp')).to.be.true;
+      expect(execKubectlStub).to.have.been.calledTwice;
+    });
+
+    it('throws instead of returning false when every probe returns no usable output', async (): Promise<void> => {
+      execKubectlStub.resolves('');
+
+      try {
+        await containerClient.hasDir('/tmp');
+        expect.fail('Expected hasDir to reject');
+      } catch (error) {
+        expect(error).to.be.instanceOf(KubeContainerOperationFailedError);
+        expect((error as KubeContainerOperationFailedError).message).to.include('hasDir /tmp');
+      }
+
+      expect(execKubectlStub).to.have.been.calledThrice;
+    });
+
+    it('falls back to /bin/sh when bash is unavailable', async (): Promise<void> => {
+      execKubectlStub
+        .onFirstCall()
+        .rejects(
+          kubectlFailure(
+            'exec failed: unable to start container process: exec: "bash": executable file not found in $PATH',
+          ),
+        );
+      execKubectlStub.onSecondCall().resolves('true');
+
+      expect(await containerClient.hasDir('/tmp')).to.be.true;
+      expect(execKubectlStub.secondCall.args[0]).to.include('/bin/sh');
+    });
   });
 });

